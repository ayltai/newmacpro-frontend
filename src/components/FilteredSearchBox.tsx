import { Check as CheckIcon, FilterAlt as FilterIcon, } from '@mui/icons-material';
import { Box, IconButton, ListItemIcon, ListItemText, Menu, MenuItem, useMediaQuery, } from '@mui/material';
import React, { ComponentType, Fragment, MouseEvent, useState, } from 'react';
import { useTranslation, } from 'react-i18next';

import { useAppDispatch, useAppSelector, } from '../hooks';
import { addFilter, removeFilter, } from '../states';
import { default as SearchBox, SearchBoxProps, } from './SearchBox';

const Filter = ({
    selectedFilters = [],
    onChange,
} : {
    selectedFilters? : string[],
    onChange         : (filter : string, selected : boolean) => void,
}) => {
    const [ anchorElement, setAnchorElement, ] = useState<null | HTMLElement>(null);

    const { t, } = useTranslation();

    const filters : Record<string, string> = t('search.filters', {
        returnObjects : true,
    });

    const handleClick = (event : MouseEvent<HTMLButtonElement>) => setAnchorElement(event.currentTarget);

    const handleClose = () => setAnchorElement(null);

    return (
        <Fragment>
            <IconButton
                sx={{
                    marginLeft : 0.5,
                }}
                size='medium'
                edge='end'
                color='inherit'
                aria-label='filter'
                onClick={handleClick}>
                <FilterIcon />
            </IconButton>
            <Menu
                open={Boolean(anchorElement)}
                anchorEl={anchorElement}
                anchorOrigin={{
                    horizontal : 'left',
                    vertical   : 'bottom',
                }}
                transformOrigin={{
                    horizontal : 'left',
                    vertical   : 'top',
                }}
                onClose={handleClose}>
                {Object.keys(filters).map(key => {
                    const handleSelect = () => {
                        handleClose();

                        onChange(key, !selectedFilters.includes(key));
                    };

                    return (
                        <MenuItem
                            key={key}
                            onClick={handleSelect}>
                            {selectedFilters.includes(key) && (
                                <>
                                    <ListItemIcon>
                                        <CheckIcon />
                                    </ListItemIcon>
                                    {filters[key]}
                                </>

                            )}
                            {!selectedFilters.includes(key) && (
                                <ListItemText inset>{filters[key]}</ListItemText>
                            )}
                        </MenuItem>
                    );
                })}
            </Menu>
        </Fragment>
    );
};

/**
 * FilteredSearchBoxes let users enter search queries and filter the results.
 * @param Component The component to wrap, normally a SearchBox component.
 */
const withFilter = (Component : ComponentType<SearchBoxProps>) => {
    const ComponentWithFilter : (props : SearchBoxProps) => React.JSX.Element = (props : SearchBoxProps) => {
        const filters  = useAppSelector(state => state.pref.filters);
        const dispatch = useAppDispatch();

        const mobileMode = useMediaQuery('(max-width: 900px)');

        const handleChange = (filter : string, selected : boolean) => {
            dispatch(selected ? addFilter(filter) : removeFilter(filter));
        };

        return (
            <Box
                display='flex'
                flexDirection='row'>
                {mobileMode && <Box flexGrow={1} />}
                <Component {...props} />
                <Filter
                    selectedFilters={filters}
                    onChange={handleChange} />
                {mobileMode && <Box flexGrow={1} />}
            </Box>
        );
    };

    return ComponentWithFilter;
};

export default withFilter(SearchBox);
