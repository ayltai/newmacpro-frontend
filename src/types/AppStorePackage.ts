export type AppStorePackage = {
    trackId                             : number,
    trackName                           : string,
    description?                        : string,
    version                             : string,
    sellerName?                         : string,
    artistName?                         : string,
    trackViewUrl?                       : string,
    averageUserRatingForCurrentVersion? : number,
    userRatingCountForCurrentVersion?   : number,
    contentAdvisoryRating?              : string,
    formattedPrice?                     : string,
    artworkUrl60?                       : string,
    artworkUrl100?                      : string,
    artworkUrl512?                      : string,
    screenshotUrls?                     : string[],
};
