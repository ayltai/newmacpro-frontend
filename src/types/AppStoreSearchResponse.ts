import type { AppStorePackage, } from './AppStorePackage';

export type AppStoreSearchResponse = {
    resultCount : number,
    results     : AppStorePackage[],
};
