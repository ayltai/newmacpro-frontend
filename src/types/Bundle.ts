import { Package, } from './Package';

export type Bundle = {
    id           : string,
    displayName? : string,
    description? : string,
    iconId?      : string,
    isPrivate?   : boolean,
    viewCount?   : number,
    createdBy?   : string,
    createdAt?   : string,
    updatedAt?   : string,
    packages     : Package[],
};
