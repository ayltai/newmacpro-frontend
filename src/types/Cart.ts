import type { Package, } from './Package';

export type Cart = {
    id       : string,
    packages : Package[],
};
