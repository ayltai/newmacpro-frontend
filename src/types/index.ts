export type { AppStorePackage, } from './AppStorePackage';
export type { AppStoreSearchResponse, } from './AppStoreSearchResponse';
export type { Bundle, } from './Bundle';
export type { Cart, } from './Cart';
export type { HomebrewCaskPackage, } from './HomebrewCaskPackage';
export type { HomebrewCorePackage, } from './HomebrewCorePackage';
export type { Package, } from './Package';
export type { Tweak, } from './Tweak';

export { SearchFilter, } from './SearchFilter';
