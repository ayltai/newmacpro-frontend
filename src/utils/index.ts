import * as Sentry from '@sentry/react';

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};

export const capitalizeFirstLetter = (input : string) => input.charAt(0).toUpperCase() + input.slice(1);
