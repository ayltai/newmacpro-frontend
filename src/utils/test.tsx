import { ThemeProvider, } from '@mui/material';
import { configureStore, Store, } from '@reduxjs/toolkit';
import { act, render, } from '@testing-library/react';
import mediaQuery from 'css-mediaquery';
import { ConfirmProvider, } from 'material-ui-confirm';
import React, { ReactNode, } from 'react';
import { Provider, } from 'react-redux';

import { bundleApi, cartApi, } from '../apis';
import { POPULAR_APPS, } from '../constants';
import { cartReducer, prefReducer, store as reduxStore, } from '../states';
import { AppTheme, } from '../styles';
import { Cart, } from '../types';

export const mockApis = async (cart? : Cart) => {
    fetchMock.doMock(request => {
        if (request.url.endsWith('/api/bundles') && request.method === 'POST') return Promise.resolve({
            status : 201,
            body   : 'Created',
        });

        if (request.url.endsWith('/api/bundles/1') && request.method === 'GET') return Promise.resolve({
            status : 200,
            body   : JSON.stringify({
                id          : '1',
                displayName : 'Dummy',
                packages    : [
                    POPULAR_APPS[0],
                ],
            }),
        });

        if (request.url.endsWith('/api/bundles/1') && request.method === 'DELETE') return Promise.resolve({
            status : 200,
            body   : 'OK',
        });

        if (request.url.endsWith('/api/carts') && request.method === 'POST') return Promise.resolve({
            status : 201,
            body   : 'Created',
        });

        return Promise.resolve({
            status : 404,
            body   : 'Not found',
        });
    });

    await act(async () => {
        await reduxStore.dispatch(bundleApi.endpoints?.getBundle.initiate({
            bundleId : '1',
            idToken  : 'idToken',
        }));
        if (cart) await reduxStore.dispatch(cartApi.endpoints.addCart.initiate(cart));
    });
};

export const createMatchMedia = (width : number) => (query : string) => ({
    matches             : mediaQuery.match(query, {
        width,
    }),
    media               : '',
    onchange            : null,
    addListener         : () => {},
    removeListener      : () => {},
    addEventListener    : () => {},
    removeEventListener : () => {},
    dispatchEvent       : () => false,
});

export const createStore = (preloadedState? : Record<string, any>) => configureStore({
    preloadedState,
    reducer    : {
        cart                      : cartReducer,
        pref                      : prefReducer,
        [ bundleApi.reducerPath ] : bundleApi.reducer,
        [ cartApi.reducerPath   ] : cartApi.reducer,
    },
    middleware : getDefaultMiddleware => getDefaultMiddleware().concat(bundleApi.middleware, cartApi.middleware),
});

export const defaultStore = createStore();

export type RootState = ReturnType<typeof defaultStore.getState>;

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : Record<string, any>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => (
        <Provider store={store}>
            <ThemeProvider theme={AppTheme}>
                <ConfirmProvider>
                    {children}
                </ConfirmProvider>
            </ThemeProvider>
        </Provider>
    ),
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
