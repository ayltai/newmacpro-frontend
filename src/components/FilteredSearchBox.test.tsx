import React from 'react';

import { SearchFilter, } from '../types';
import { fireEvent, render, waitFor, } from '../utils/test';

import FilteredSearchBox from './FilteredSearchBox';

const mockUseDispatch = jest.fn();

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useDispatch : () => mockUseDispatch,
}));

describe('<FilteredSearchBox />', () => {
    beforeEach(() => mockUseDispatch.mockClear());

    it('renders correctly', () => expect(render(<FilteredSearchBox />).asFragment()).toMatchSnapshot());

    it('adds filter when filter is selected', async () => {
        const { getByRole, } = render(<FilteredSearchBox />);

        fireEvent.click(getByRole('button'));

        expect(getByRole('menu')).toBeInTheDocument();

        fireEvent.click(getByRole('menuitem', {
            name : 'Include apps',
        }));

        await waitFor(() => {
            expect(mockUseDispatch).toHaveBeenCalledTimes(1);
            expect(mockUseDispatch).toHaveBeenCalledWith({
                type    : 'pref/removeFilter',
                payload : SearchFilter.Apps,
            });
        });
    });
});
