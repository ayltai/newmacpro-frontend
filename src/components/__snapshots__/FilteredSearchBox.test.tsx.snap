// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilteredSearchBox /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1xhj18k"
  >
    <div
      class="css-1kjapbk"
    >
      <div
        class="css-1iz4iln"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </div>
      <div
        class="MuiInputBase-root MuiInputBase-colorPrimary css-4pgk2p-MuiInputBase-root"
      >
        <input
          class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      aria-label="filter"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-vyhsqy-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="FilterAltIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M4.25 5.61C6.27 8.2 10 13 10 13v6c0 .55.45 1 1 1h2c.55 0 1-.45 1-1v-6s3.72-4.8 5.74-7.39c.51-.66.04-1.61-.79-1.61H5.04c-.83 0-1.3.95-.79 1.61z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
